rllib/agents/impala/vtrace_tf_policy.py

```diff
@@ -158,197 +167,273 @@ def _make_time_major(policy, seq_lens, tensor, drop_last=False):
-    if policy.config["_tf_policy_handles_more_than_one_loss"]:
-        optimizers = force_list(optimizer)
-        losses = force_list(loss)
-        assert len(optimizers) == len(losses)
-        clipped_grads_and_vars = []
-        for optim, loss_ in zip(optimizers, losses):
-            grads_and_vars = optim.compute_gradients(
-                loss_, policy.model.trainable_variables()
+    Returns:
+        A TF Policy to be used with ImpalaTrainer.
+    """
+    # VTrace mixins are placed in front of more general mixins to make sure
+    # their functions like optimizer() overrides all the other implementations
+    # (e.g., LearningRateSchedule.optimizer())
+    class VTraceTFPolicy(
+        VTraceClipGradients,
+        VTraceOptimizer,
+        LearningRateSchedule,
+        EntropyCoeffSchedule,
+        base,
+    ):
+        def __init__(
+            self,
+            obs_space,
+            action_space,
+            config,
+            existing_model=None,
+            existing_inputs=None,
+        ):
+            # First thing first, enable eager execution if necessary.
+            base.enable_eager_execution_if_necessary()
+
+            config = dict(ray.rllib.agents.impala.impala.DEFAULT_CONFIG, **config)
+
+            # Initialize base class.
+            base.__init__(
+                self,
+                obs_space,
+                action_space,
+                config,
+                existing_inputs=existing_inputs,
+                existing_model=existing_model,
             )
-            clipped_g_and_v = []
-            for g, v in grads_and_vars:
-                if g is not None:
-                    clipped_g, _ = tf.clip_by_global_norm(
-                        [g], policy.config["grad_clip"]
-                    )
-                    clipped_g_and_v.append((clipped_g[0], v))
-            clipped_grads_and_vars.append(clipped_g_and_v)
-
-        policy.grads = [g for g_and_v in clipped_grads_and_vars for (g, v) in g_and_v]
-    # Only one optimizer and and loss term.
-    else:
-        grads_and_vars = optimizer.compute_gradients(
-            loss, policy.model.trainable_variables()
-        )
-        grads = [g for (g, v) in grads_and_vars]
-        policy.grads, _ = tf.clip_by_global_norm(grads, policy.config["grad_clip"])
-        clipped_grads_and_vars = list(
-            zip(policy.grads, policy.model.trainable_variables())
-        )
 
-    return clipped_grads_and_vars
+            VTraceClipGradients.__init__(self)
+            VTraceOptimizer.__init__(self)
+            LearningRateSchedule.__init__(self, config["lr"], config["lr_schedule"])
+            EntropyCoeffSchedule.__init__(
+                self, config["entropy_coeff"], config["entropy_coeff_schedule"]
+            )
 
+            # Note: this is a bit ugly, but loss and optimizer initialization must
+            # happen after all the MixIns are initialized.
+            self.maybe_initialize_optimizer_and_loss()
```

Comment:
is there a reason why this is called maybe... does that imply that it doesnt happen always, and if so, why?

Reply:
TF2 doesn't need to initialize these. You only need to construct loss graphs for static graph TF1.
TF2 takes the chance to initialize a few things like view_requirements I think.